apps/hubble/src/addon/src/store/link_store.rs

```
@@ -0,0 +1,675 @@
+        if link_body.target.is_some() && (link_body.r#type.is_empty() || link_body.r#type.len() == 0) {
+            return Err(HubError::validation_failure("targetID provided without type"));
+        }
+
+        if !link_body.r#type.is_empty()
+            && (link_body.r#type.len() > Self::LINK_TYPE_BYTE_SIZE || link_body.r#type.len() == 0) {
+            return Err(HubError::validation_failure("link type invalid - non-empty link type found with invalid length"));
+        }
+
+        let mut key = Vec::with_capacity(
+            Self::ROOT_PREFIXED_FID_BYTE_SIZE + Self::POSTFIX_BYTE_SIZE + Self::LINK_TYPE_BYTE_SIZE + Self::TARGET_ID_BYTE_SIZE
+        );
+
+        // TODO: does the fid and rtype need to be padded? Is it okay not the check their lengths?
+        key.extend_from_slice(&make_user_key(fid));
+        key.push(UserPostfix::LinkRemoves.as_u8());
+        key.extend_from_slice(&link_body.r#type.as_bytes());
+        match link_body.target {
+            None => {}
+            Some(Target::TargetFid(fid)) => {
+                key.extend_from_slice(&make_fid_key(fid as u32)[..Self::TARGET_ID_BYTE_SIZE])
+            }
+        }
+
+        Ok(key)
+    }
+
+    /// Generates a unique key used to store a LinkAdd Message in the LinksByTargetAndType index.
+    /// Returns RocksDB index key of the form <RootPrefix>:<target_key>:<fid?>:<tsHash?>
+    ///
+    /// # Arguments
+    /// * `target` - target ID of the object being reacted to (currently just cast id)
+    /// * `fid` - the fid of the user who created the link
+    /// * `ts_hash` - the timestamp hash of the link message
+    fn links_by_target_key(
+        target: &Target,
+        fid: u32,
+        ts_hash: Option<&[u8; TS_HASH_LENGTH]>,
+    ) -> Vec<u8> {
+        if fid != 0 && (ts_hash.is_none() || ts_hash.is_some_and(|tsh| tsh.len() == 0)) {
+            panic!("bad_request.validation_failure - fid provided without timestamp hash");
```

I don't think we should panic in these functions?  We should return an Err with the right error code and message.

---

fair enough, updated